File: src/pysmesh/unify.py
"""Same-domain topology healing (Tier-2).

Public surface: :class:`UnifyParams`, :class:`UnifyResult`, and :func:`unify_same_domain`.
These wrap the low-level ``_core.unify_same_domain`` (which returns raw BREP bytes + NumPy
arrays) in frozen dataclasses and validate parameters up front.

``unify_same_domain`` merges adjacent faces that lie on one underlying surface — and
collinear edges — into a single face/edge, deleting the artificial seams that
over-segmented STEP imports carry. It is a pure OCCT B-rep operation
(``ShapeUpgrade_UnifySameDomain``): the seam face/edge is removed from the topology, so a
downstream mesher never places nodes along it. Both the input and the healed output cross
the boundary as raw BREP bytes, exactly like :func:`load_brep`.

``face_map`` / ``edge_map`` follow pySMESH's 1-based TopExp id convention (the same ids
:meth:`Shape.faces` / :meth:`Shape.edges` return): ``face_map[i - 1]`` is the new face id
that original face id ``i`` survives as, or ``-1`` if it was removed. Merged originals share
one survivor (many-to-one), which is how a consumer composes a pre-heal tag onto the
post-heal geometry.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import cast

import numpy as np
from numpy.typing import NDArray

from ._core import PysmeshError
from ._core import unify_same_domain as _unify_same_domain

# OCCT ShapeUpgrade_UnifySameDomain defaults (Precision::Confusion for the linear tolerance).
# The angular default is Precision::Angular (~1e-12 rad); passing 0.0 lets OCCT clamp to it.
_DEFAULT_LINEAR_TOL: float = 1.0e-7


@dataclass(frozen=True)
class UnifyParams:
    """Same-domain healing parameters.

    Attributes:
        unify_faces: Merge adjacent faces sharing one underlying surface.
        unify_edges: Merge collinear/cotangent edges sharing one underlying curve.
        concat_bsplines: Concatenate merged B-spline geometry into a single surface/curve
            instead of leaving it trimmed. Off by default (structure-preserving).
        linear_tol: Chord tolerance [model units] for deciding coplanarity (> 0).
        angular_tol_deg: Maximum connection angle [degrees] for merging; two shapes forming
            a sharper angle are left separate. 0.0 uses OCCT's tight default (~1e-12 rad).

    Raises:
        PysmeshError: On any out-of-range parameter.
    """

    unify_faces: bool = True
    unify_edges: bool = True
    concat_bsplines: bool = False
    linear_tol: float = _DEFAULT_LINEAR_TOL
    angular_tol_deg: float = 0.0

    def __post_init__(self) -> None:
        if not self.linear_tol > 0.0:
            raise PysmeshError(
                f"UnifyParams.linear_tol must be > 0 (got {self.linear_tol})."
            )
        if self.angular_tol_deg < 0.0:
            raise PysmeshError(
                f"UnifyParams.angular_tol_deg must be >= 0 (got {self.angular_tol_deg})."
            )
        if not (self.unify_faces or self.unify_edges):
            raise PysmeshError(
                "UnifyParams must enable at least one of unify_faces / unify_edges."
            )


@dataclass(frozen=True)
class UnifyResult:
    """Result of :func:`unify_same_domain`.

    Attributes:
        brep: The healed shape as BREP bytes (re-loadable via :func:`load_brep`).
        n_faces_before: Face count of the input shape.
        n_faces_after: Face count of the healed shape (<= ``n_faces_before``).
        n_edges_before: Edge count of the input shape.
        n_edges_after: Edge count of the healed shape (<= ``n_edges_before``).
        face_map: (n_faces_before,) int32 — new 1-based face id per original face id (row
            ``i`` is original id ``i + 1``), or ``-1`` if the face was removed.
        edge_map: (n_edges_before,) int32 — same mapping for edges.
    """

    brep: bytes
    n_faces_before: int
    n_faces_after: int
    n_edges_before: int
    n_edges_after: int
    face_map: NDArray[np.int32]
    edge_map: NDArray[np.int32]


def unify_same_domain(brep: bytes, params: UnifyParams | None = None) -> UnifyResult:
    """Merge same-domain faces/edges of a BREP shape.

    Args:
        brep: Input shape as BREP bytes (e.g. from :func:`load_brep`'s source, or any OCCT
            ``BRepTools::Write`` output).
        params: Healing parameters. Defaults to :class:`UnifyParams` (faces + edges, OCCT
            default tolerances).

    Returns:
        The healed shape plus before/after counts and the id-composition maps.

    Raises:
        PysmeshError: On a malformed BREP, or if the healed shape fails OCCT's validity
            check (``BRepCheck_Analyzer``).
    """
    p = params if params is not None else UnifyParams()
    raw = _unify_same_domain(
        brep,
        p.unify_faces,
        p.unify_edges,
        p.concat_bsplines,
        p.linear_tol,
        math.radians(p.angular_tol_deg),
    )
    return UnifyResult(
        brep=cast("bytes", raw["brep"]),
        n_faces_before=cast("int", raw["n_faces_before"]),
        n_faces_after=cast("int", raw["n_faces_after"]),
        n_edges_before=cast("int", raw["n_edges_before"]),
        n_edges_after=cast("int", raw["n_edges_after"]),
        face_map=cast("NDArray[np.int32]", raw["face_map"]),
        edge_map=cast("NDArray[np.int32]", raw["edge_map"]),
    )

File: src/bindings/unify.cpp
// pySMESH binding — topology healing: unify_same_domain.
//
// Wraps OCCT's ShapeUpgrade_UnifySameDomain (TKShHealing) — a pure B-rep operation that
// merges adjacent faces lying on the same underlying surface (and collinear edges) into a
// single face/edge, removing the artificial seams that over-segmented STEP imports carry.
// Unlike a mesher's "compound" hint, this is real geometric healing: the seam face/edge is
// deleted from the topology, so a downstream mesher never places nodes along it.
//
// No SMESH here — this is OCCT-only and shares just the BREP bytes bridge, ShapeData index
// convention, and PysmeshError with the rest of pysmesh. Input and output both cross the
// boundary as raw BREP bytes (BRepTools::Read/Write), exactly like load_brep.
//
// Index convention (matches Shape.faces()/edges() and flux's tag composition): face/edge
// ids are 1-based TopExp::MapShapes ordinals with a per-kind type filter. face_map[i-1] is
// the new 1-based face id that original face id i survives as, or -1 if it was removed;
// edge_map is the same for edges. Merged originals share one survivor (many-to-one).

#include <cstdint>
#include <sstream>
#include <string>
#include <vector>

#include <BRepCheck_Analyzer.hxx>
#include <BRepTools.hxx>
#include <BRepTools_History.hxx>
#include <BRep_Builder.hxx>
#include <ShapeUpgrade_UnifySameDomain.hxx>
#include <Standard_Handle.hxx>
#include <TopAbs_ShapeEnum.hxx>
#include <TopExp.hxx>
#include <TopTools_IndexedMapOfShape.hxx>
#include <TopTools_ListOfShape.hxx>
#include <TopoDS_Shape.hxx>

#include "common.hpp"

namespace pysmesh {
namespace {

// Read a BREP shape from in-memory bytes (mirrors shape.cpp::load_brep, which is file-local
// there). Raises on parse failure or a null result.
TopoDS_Shape read_brep(const py::bytes& data) {
  const std::string buffer = data;
  std::istringstream stream(buffer);
  TopoDS_Shape shape;
  BRep_Builder builder;
  try {
    BRepTools::Read(shape, stream, builder);
  } catch (const std::exception& e) {
    throw PysmeshError(std::string("BREP read failed: ") + e.what());
  }
  if (shape.IsNull()) {
    throw PysmeshError("BREP read produced a null shape (empty or malformed data)");
  }
  return shape;
}

// Serialize a shape back to BREP bytes.
py::bytes write_brep(const TopoDS_Shape& shape) {
  std::ostringstream stream;
  try {
    BRepTools::Write(shape, stream);
  } catch (const std::exception& e) {
    throw PysmeshError(std::string("BREP write failed: ") + e.what());
  }
  return py::bytes(stream.str());
}

// Map every original sub-shape (1-based in old_map) to its surviving 1-based id in new_map:
//   removed          -> -1
//   modified/merged  -> the (first) survivor recorded in history, looked up in new_map
//   unchanged        -> itself (IsSame) in new_map
// A survivor that cannot be located in the unified shape is a history/topology inconsistency
// and raises rather than silently emitting a wrong id.
py::array_t<std::int32_t> build_map(const TopTools_IndexedMapOfShape& old_map,
                                    const TopTools_IndexedMapOfShape& new_map,
                                    const Handle(BRepTools_History) & hist, const char* kind) {
  const int n = old_map.Extent();
  py::array_t<std::int32_t> out(static_cast<py::ssize_t>(n));
  std::int32_t* d = out.mutable_data();
  for (int i = 1; i <= n; ++i) {
    const TopoDS_Shape& s = old_map.FindKey(i);
    if (!hist.IsNull() && hist->IsRemoved(s)) {
      d[i - 1] = -1;
      continue;
    }
    TopoDS_Shape survivor = s;
    if (!hist.IsNull()) {
      const TopTools_ListOfShape& mod = hist->Modified(s);
      if (!mod.IsEmpty()) {
        survivor = mod.First();
      }
    }
    const int idx = new_map.FindIndex(survivor);
    if (idx == 0) {
      throw PysmeshError(std::string("unify_same_domain: original ") + kind + " id " +
                          std::to_string(i) +
                          " could not be mapped to the unified shape "
                          "(history/topology inconsistency).");
    }
    d[i - 1] = static_cast<std::int32_t>(idx);
  }
  return out;
}

py::dict unify_same_domain(const py::bytes& brep, bool unify_faces, bool unify_edges,
                            bool concat_bsplines, double linear_tol, double angular_tol_rad) {
  const TopoDS_Shape shape = read_brep(brep);

  // Original sub-shape maps (source of the 1-based ids), captured before Build().
  TopTools_IndexedMapOfShape old_faces;
  TopTools_IndexedMapOfShape old_edges;
  TopExp::MapShapes(shape, TopAbs_FACE, old_faces);
  TopExp::MapShapes(shape, TopAbs_EDGE, old_edges);

  // ctor arg order is (shape, UnifyEdges, UnifyFaces, ConcatBSplines).
  ShapeUpgrade_UnifySameDomain unifier(shape, unify_edges, unify_faces, concat_bsplines);
  unifier.SetLinearTolerance(linear_tol);
  unifier.SetAngularTolerance(angular_tol_rad);
  // Safe mode: operate on a copy so the input sub-shape maps above stay identity-stable and
  // the history bridges old -> new cleanly.
  unifier.SetSafeInputMode(true);

  TopoDS_Shape result;
  {
    py::gil_scoped_release release;
    try {
      unifier.Build();
    } catch (const std::exception& e) {
      py::gil_scoped_acquire acquire;
      throw PysmeshError(std::string("ShapeUpgrade_UnifySameDomain::Build failed: ") +
                          e.what());
    }
    result = unifier.Shape();
  }

  if (result.IsNull()) {
    throw PysmeshError("unify_same_domain produced a null shape.");
  }
  // Fail loud on an invalid healed shape rather than shipping a corrupt BREP downstream.
  if (!BRepCheck_Analyzer(result).IsValid()) {
    throw PysmeshError("unify_same_domain produced an invalid shape "
                        "(BRepCheck_Analyzer reported errors).");
  }

  TopTools_IndexedMapOfShape new_faces;
  TopTools_IndexedMapOfShape new_edges;
  TopExp::MapShapes(result, TopAbs_FACE, new_faces);
  TopExp::MapShapes(result, TopAbs_EDGE, new_edges);

  const Handle(BRepTools_History) hist = unifier.History();

  py::dict out;
  out["brep"] = write_brep(result);
  out["n_faces_before"] = old_faces.Extent();
  out["n_faces_after"] = new_faces.Extent();
  out["n_edges_before"] = old_edges.Extent();
  out["n_edges_after"] = new_edges.Extent();
  out["face_map"] = build_map(old_faces, new_faces, hist, "face");
  out["edge_map"] = build_map(old_edges, new_edges, hist, "edge");
  return out;
}

}  // namespace

void bind_unify(py::module_& m) {
  m.def("unify_same_domain", &unify_same_domain, py::arg("brep"), py::arg("unify_faces"),
        py::arg("unify_edges"), py::arg("concat_bsplines"), py::arg("linear_tol"),
        py::arg("angular_tol_rad"),
        "Merge same-domain faces/edges of a BREP shape (OCCT ShapeUpgrade_UnifySameDomain). "
        "Returns a dict: unified BREP bytes, before/after face+edge counts, and int32 "
        "face_map/edge_map (old 1-based id -> new 1-based id, -1 if removed). Low-level: the "
        "pysmesh.unify_same_domain wrapper builds UnifyParams/UnifyResult around it.");
}

}  // namespace pysmesh
